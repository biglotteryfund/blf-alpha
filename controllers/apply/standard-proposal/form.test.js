/* eslint-env jest */
'use strict';
const omit = require('lodash/omit');

const { mockResponse } = require('./mocks');
const formBuilder = require('./form');

function mapMessages(validationResult) {
    return validationResult.messages.map((detail) => detail.msg);
}

test('empty form', () => {
    const result = formBuilder().validation;
    expect(mapMessages(result)).toMatchSnapshot();
});

test('valid form', () => {
    const data = mockResponse();
    const result = formBuilder({ data }).validation;
    expect(result.error).toBeUndefined();

    expect(result.value).toMatchSnapshot({
        yourIdeaProject: expect.any(String),
        yourIdeaCommunity: expect.any(String),
        yourIdeaActivities: expect.any(String),
    });
});

test('invalid form', () => {
    const form = formBuilder();

    const result = form.validate({
        projectCountries: 'invalid-country',
        projectLocation: null,
        projectCosts: 10000,
        projectDurationYears: 10,
        organisationLegalName: 'Same organisation name',
        organisationTradingName: 'Same organisation name',
    });

    expect(mapMessages(result)).toMatchSnapshot();
});

test('require region when england is selected', function () {
    const data = omit(mockResponse(), 'projectRegions');
    const form = formBuilder({ data });

    expect(mapMessages(form.validation)).toEqual(
        expect.arrayContaining(['Select one or more regions'])
    );
});

test('strip region outside of england', function () {
    const data = mockResponse({
        projectCountries: ['northern-ireland'],
        projectRegions: ['midlands', 'north-west'],
    });

    const form = formBuilder({ data });

    expect(form.validation.value).not.toHaveProperty('projectRegions');
});

test('strip other region selections when all-england is selected', function () {
    const form = formBuilder({
        data: mockResponse({
            projectCountries: ['england'],
            projectRegions: ['all-england', 'midlands', 'north-west'],
        }),
    });

    expect(form.validation.value.projectRegions).toEqual(['all-england']);
});

test('strip projectLocation when applying for more than one country', () => {
    const form = formBuilder({
        data: mockResponse({
            projectCountries: ['england', 'scotland'],
            projectLocation: 'this-should-be-stripped',
        }),
    });

    expect(form.validation.value).not.toHaveProperty('projectLocation');
});

test('strip projectDurationYears when applying for more than one country', () => {
    const form = formBuilder({
        data: mockResponse({
            projectCountries: ['england', 'scotland'],
            projectDurationYears: 5,
        }),
    });

    expect(form.validation.value).not.toHaveProperty('projectDurationYears');
});

test('set projectDurationYears to 1 when applying for England', () => {
    const form = formBuilder({
        data: mockResponse({
            projectCountries: ['england'],
            projectDurationYears: 5,
        }),
    });
    const salesforceResult = form.forSalesforce();
    expect(salesforceResult.projectDurationYears).toBe(1);
});

test('organisation sub-type required for statutory-body', function () {
    const requiredData = mockResponse({ organisationType: 'statutory-body' });
    const requiredResult = formBuilder({ data: requiredData }).validation;

    expect(mapMessages(requiredResult)).toEqual(
        expect.arrayContaining(['Tell us what type of statutory body you are'])
    );

    const validData = mockResponse({
        organisationType: 'statutory-body',
        organisationSubType: 'fire-service',
    });
    const result = formBuilder({ data: validData }).validation;
    expect(result.error).toBeUndefined();
});

test('language preference required in wales', function () {
    const form = formBuilder({
        data: mockResponse({
            projectCountries: ['england', 'wales'],
        }),
    });

    expect(mapMessages(form.validation)).toEqual(
        expect.arrayContaining([expect.stringContaining('Select a language')])
    );

    const formValid = formBuilder({
        data: mockResponse({
            projectCountries: ['england', 'wales'],
            contactLanguagePreference: 'welsh',
        }),
    });

    expect(formValid.validation.error).toBeUndefined();

    const formStrip = formBuilder({
        data: mockResponse({
            projectCountries: ['england'],
            contactLanguagePreference: 'welsh',
        }),
    });

    expect(formStrip.validation.value).not.toHaveProperty(
        'contactLanguagePreference'
    );
});

test.each([
    'organisationTradingName',
    'contactPhone',
    'contactCommunicationNeeds',
])('optional %p field', function (fieldName) {
    const data = mockResponse();
    const form = formBuilder({ data });

    const expected = omit(data, fieldName);
    const result = form.validate(expected);
    expect(result.error).toBeUndefined();
});
