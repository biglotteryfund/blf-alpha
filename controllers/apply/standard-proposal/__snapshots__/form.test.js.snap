// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty form 1`] = `
Array [
  "Enter a project name",
  "Select a country",
  "Select one or more regions",
  "Tell us all the locations that you'll be running your project in",
  "Enter a postcode",
  "Tell us what you would like to do",
  "Tell us how your project involves your community",
  "Tell us how your idea fits in with other local activities",
  "Enter the full legal name of the organisation",
  "Select yes or no.",
  "Enter a full UK address",
  "Select a type of organisation",
  "Enter first and last name",
  "Enter an email address",
  "Check this box to show you understand",
  "Select an option",
  "Select the specific group(s) of people your leadership identifies in",
]
`;

exports[`invalid form 1`] = `
Array [
  "Enter a project name",
  "Select a country",
  "Select a location",
  "Tell us all the locations that you'll be running your project in",
  "Enter a postcode",
  "The amount you ask for must be more than £10,000. If you need less than this,  <a href=\\"/funding/under10k\\">you can apply for under £10,000 here</a>.",
  "Select a project length",
  "Tell us what you would like to do",
  "Tell us how your project involves your community",
  "Tell us how your idea fits in with other local activities",
  "Select yes or no.",
  "Enter a full UK address",
  "Select a type of organisation",
  "Enter first and last name",
  "Enter an email address",
  "Check this box to show you understand",
  "Select an option",
  "Select the specific group(s) of people your leadership identifies in",
]
`;
